### ISSUE TITLE
feature: `--foo+=mypath` to append to `foo`

### ISSUE THREAD

**Issue by timotheecour** [c-blake/cligen#17]
this is a common use case:
supporting appending to an array as opposed to resetting.

## example 1: append to a seq[string]

```
proc foobar(imports = @[".", ".."], libs = @["dl", "pthread"], paths: seq[string]) = discard
```

The following rules are very intuitive and add lots of flexibility, especially for letting users choose how to override default configs:
```
./app
imports=@[".", ".."]

./app --imports="my other lib"
imports=@["my other lib"]

./app --imports= rest of args #notice space after `=`
imports=@[""]

./app --imports+="baz"
imports=@[".", "..", "baz"]

./app --imports+="baz1" --imports+="baz2"
imports=@[".", "..", "baz1", "baz2"]

./app --imports="baz1" --imports+="baz2"
imports=@["baz1", "baz2"]

./app --imports+= rest of args  #notice space after `=`
imports=@[".", "..", ""]
```

## example 2: append to a string

same behavior as above, eg:
```
proc foobar(path = "foo:bar", paths: seq[string]) = discard

./app --path+=":baz"
imports="foo:bar:baz"
```

## example 3: append to a seq[int]
same concept...

## NOTE
`&=` would be more Nim-esque, but that is inconvenient for shells as `&` would have to be escaped.

**Comment by c-blake** [comment]
I don't really have this in the documentation yet, and it isn't _quite_ what you're asking for, but you should look at https://github.com/c-blake/cligen/blob/master/test/FancyRepeats.nim, in particular the ``'s'/"stuff"`` option key.  You can at least append up a ``seq`` with repeated utterances of the option key.  The ``'v'`` option there shows how to implement an accumulating "NoVal" style argument like a verbosity level with ``"-vvvv"``. A similar approach would work for strings, but it might make sense to have a distinct string type there since non-appending strings are more common.

One would need more syntax, perhaps along the lines you suggest, to distinguish starting over from empty vs. appending, though, which is kind of in the same dimension as your OptionalNoVal idea.

**Comment by timotheecour** [comment]
* I really prefer `-v:3` to `-vvv` (especially in case a wrapper script already set `-vv` for eg, in which case it would be come -vvvvv which is probably not what user wants)
* I think `--foo=bar` and `--foo+=bar` covers all the use cases you mention, and works with both strings and arrays (eg arrays of strings or ints) with no ambiguity

**Comment by c-blake** [comment]
So, you prefer a ``verb: int=0`` or whatever parameter and your commands can all use that.  That's fine.  You can even just type ``-v3``.  No problem there, and that all works now.  (That also solves your OptionalNoVal issue, if you just follow it more generally.)  Not everyone will share that preference, though.  So, no guarantees about what commands by others that use this library will wind up throwing your way.

As to ``+=``, I will look at that when I'm looking at the OptionalNoVal idea.  It's kinda related..just making sure there are no edge case gotchas/implementing the logic.

**Comment by c-blake** [comment]
I don't think this ``+=`` syntax is really worth it.  It's beyond common syntax for CLIs.  Also, besides appending/augmenting, one might want subtracting or prepending or other sorts of manipulations.  It's kind of a slippery slope.  I thought about maybe letting any non-alphanumeric be a ``sepChar`` in ``parseopt3`` and having the ``getopt`` yield a 4-tuple of ("kind", "sep", "key", "val") instead of the current 3-tuple.  Then the "sep" could be passed down through to ``argParse`` and CLI authors could endow different meanings for '+' and '-' and '/' and so forth by defining their own ``argParse()``s.  However, that tuple shift would really make ``parseopt3`` no longer compatible with ``parseopt``..a non-trivial cost for uncommon syntax.

Also, at present, ``cligen`` does not even have a builtin "string -> seq[T]" conversion.  Any such beast needs at a minimum a delimiting convention, but possibly also quoting.  Quoting could probably be avoided by the user self-delimiting - say the first char of the string is always the delimiter as in ``":A:B:C"`` parses to ``[ "A", "B", "C"]`` where the first character **always** has to be the delimiter.  Even this choice is beyond the realm of no surprises/common syntax, though.

Since that "string -> seq[]" syntax already needs to be chosen, and there is no standard for it, it could as easily be that a ``"+"`` character there can mean to append, as in ``--foo=+:A:B:C``.  This is possible right now with no mods.  The basic user-defined work for it was/is already needed for a ``seq[string]`` to be received in a single parameter after a short or long option key ``'='``.  The other use cases of ``string``-concatenation or ``int``-adjustment seem pretty contrived/unusual, to be honest.

**Comment by timotheecour** [comment]
hi, no, havent lost interest, just got caught up with other stuff. thats cool, one thing that comes to mind though is ambiguity of =+
how to differentiate bw assigning a string +bar vs appending bar to an array of strings.

+= didnt have that ambiguity

**Comment by c-blake** [comment]
Right.  Well, to allow users to reset to a sequence starting with ``"+bar"`` we probably need a 3rd mutation mode tag.  I would say ``=`` is most natural. Then to assign an element starting with a tag ``-s[maybe<SEPCHAR>][maybe=]<TAG>bar``, e.g. or ``-s:=+bar`` or ``-s<SPACE>=-bar``.  In terms of scripts where the nature of the string may be unknown, the ``'='`` mode tag is always allowed, but not required.  So, ``-s:=$p`` would work as expected to reset to whatever the shell expands ``$p`` to, and similarly for ``-s:+$p``, ``-s:-$p`` and so on. { Using ``=`` instead of ``:`` here as a parseopt3 sepChar works, but the multiple meanings of ``=`` in context make it harder to read. }  I went ahead an implemented this in my most recent commits.

Of course, this means there is one more initial character that creates a need for special treatment, namely ``'='``.  And this all is more to explain to CLI users.  This whole aggregate editing scenario is more to explain, though.  In the wider world, repetition of the option key is what you use to append (as in ``cc -Dmac1 -Ipath1 -Dmac2 -Ipath2 ...`` like in ``test/FancyRepeats.nim`` as I mentioned earlier).  That repetition command syntax does not have these ambiguities (or even a need for a sequence element delimiting convention), but also does not have reset or deletion powers.  More powers requires more knowledge and all that.  Maybe this is all "too much power".

**Comment by c-blake** [comment]
Ok.  I tried using commands with my above proposed syntax and I didn't like all the levels of syntax.  So, I figured out a way to retain drop-in compatibility with the stdlib parseopt (ditching the getopt iterator, basically) while adding the original new syntax you asked for.  I also added a few extras.  E.g., there is also a "prepend mode" via "^=".  Things work for strings and seqs (and sets).  I'm not sure if number types should support ``[+-*/]=``.  Maybe.
